### ISSUE TITLE
[8 weeks] Add Animal Groups

### ISSUE THREAD

**Issue by BradleyChatha** [Aim-Educational/FarmMaster#14]
Allow groups of animals to be created:

* Groups are given a name and description. [Day 1]
* Animals can belong to multiple groups. [Day 1]
* When creating or editing an animal, there should be a dropdown to assign it to a group. [Day 1]
* Similarly, when editing a group there should be a mini-editor to easily add/remove animals to and from the group. [Days 2 and 3, if we go with this]
* Ideally there should also be two specialised tools, one for manual mass assignment/deassignment, and one for automated mass assignment. [Days 4 to 8]

Manual mass assignment:
* Things like "Add/Remove all animals born between two dates"; "Add only Cows"; "Add anything *but* Cows", etc.
* Allow these rulesets to be composed together before performing the action.

Automatic mass assignment:
* Similar to manual assignment, where you can compose different rules together (multiple rulesets should be able to be composed for a single group).
* When a new ruleset is added, all current animals (that aren't dead/sold/whatever) will be checked to see if they match the ruleset, and if they do, then they are added. (Or maybe we have a button for this)
* After an animal is created, all rulesets from all groups will be checked to see if it should be added to these groups.
* A complication occurs when an automatic ruleset adds an animal, but it is manually removed from the group. There needs to be a form of tracking in place to prevent the animal from being auto-added again. The UI should also list these animals being blocked from being added.
* Accidental or malformed rulesets can be undone via the manual mass unassignment, it's just a bit annoying.

Other(Days 9 to 14):
* Once an animal reaches 'end of system' (dies, sold, out of our care for a while, etc.) then it should be removed from all of it's groups automatically. Even better would to add a toggle option on groups on whether this behaviour should be performed or not.
* Groups can either be archived and/or deleted, depending on needs. Archiving prevents any animals from being added or removed from the group, which also means automatic rulesets stop working. Any other various functions that can be performed on groups should also stop functioning.
* Archived groups should be hidden behind a checkbox on it's index page. It shouldn't show up at all in dropdowns. It should get special styling everywhere else to indicate that it is no longer useable.
* Unarchiving could be desirable, but is not a priority for release.
* Life event entries should be able to be applied to all animals in a group. This should actually be really easy... *should* with the way I've designed life events.

**Comment by BradleyChatha** [comment]
Don't ask. Just don't:

```
SELECT "Animals".*, "MapLifeEventEntryToAnimals".*, "LifeEventEntries".*, "LifeEvents".*, "LifeEventDynamicFieldValues".*, "LifeEventDynamicFieldInfo".*
FROM "Animals"
INNER JOIN "MapLifeEventEntryToAnimals"
	ON "MapLifeEventEntryToAnimals"."AnimalId" = "Animals"."AnimalId"
	INNER JOIN "LifeEventEntries"
		ON "LifeEventEntries"."LifeEventEntryId" = "MapLifeEventEntryToAnimals"."LifeEventEntryId"
		INNER JOIN "LifeEvents"
			ON "LifeEvents"."LifeEventId" = "LifeEventEntries"."LifeEventId"
		INNER JOIN "LifeEventDynamicFieldValues" 
			ON "LifeEventDynamicFieldValues"."LifeEventEntryId" = "LifeEventEntries"."LifeEventEntryId"
			INNER JOIN "LifeEventDynamicFieldInfo"
				ON "LifeEventDynamicFieldInfo"."LifeEventDynamicFieldInfoId" = "LifeEventDynamicFieldValues"."LifeEventDynamicFieldInfoId"
WHERE
	((    "LifeEvents"."Name" = 'Born' 
	  AND "LifeEvents"."Flags" & 2 = 2 
	  AND "LifeEventEntries"."LifeEventId" = "LifeEvents"."LifeEventId"
	  AND "LifeEventDynamicFieldValues"."LifeEventEntryId" = "LifeEventEntries"."LifeEventEntryId"
	  AND ("LifeEventDynamicFieldValues"."Value"::json->>'v')::timestamp with time zone < '01/01/2020'::timestamp
	 ))
;
```

**Comment by BradleyChatha** [comment]
tbh I just need to finish the UI for the list of automated scripts, and then this'll be just barely useable. Any other QoL addtions or tweaks or whatever will come from Andy's feedback.

**Comment by BradleyChatha** [comment]
Closed by 061399b3e12497dfffa2951ba149fa22c205e4ce
